agentspawner/daemon.py

```diff
@@ -0,0 +1,126 @@
+#! /bin/python3
+
+"""
+Manage ideal number of "agent" like resources in Resalloc.
+"""
+
+import subprocess
+import time
+import logging
+
+from resalloc.client import (
+    Connection as ResallocConnection,
+    Ticket,
+)
+
+RESALLOC_SERVER = "http://localhost:49100"
+
+
+class SpawnerPool:
+    """
+    Manage ideal number of "agent" like resources in Resalloc.
+    """
+    sleep = 30
+
+    def __init__(self, resalloc_connection, logger):
+        self.tags = ["A"]
+        # TODO: use a persistent storage so we can restart the process
+        self.tickets = []
+        self.conn = resalloc_connection
+        self.log = logger
+
+    def call_converge_to(self):
+        """ Execute the configured hook script """
+        while True:
+            result = subprocess.run(["./hook-converge-to"], capture_output=True,
+                                    check=False)
+            if result.returncode == 0:
+                try:
+                    return int(result.stdout.decode("utf-8").strip())
+                except ValueError:
+                    pass
+
+            self.log.debug("Failing to run converge-to hook")
+
+    def call_take(self, data):
+        """
+        Call hook that prepares the resource
+        """
+        return not subprocess.run(["./hook-take", f"{data}"], check=True)
+
+    def call_release(self, data):
+        """
+        Call hook that releases the resource
+        """
+        result = subprocess.run(["./hook-release", f"{data}"], check=False)
+        return not result.returncode
+
+    def start(self, count):
+        """ Start N agent-like resources """
+        self.log.info("Starting %s resources", count)
+        for _ in range(count):
+            ticket = self.conn.newTicket(self.tags)
+            self.log.debug("Taking ticket id %s", ticket.id)
+            self.tickets.append(ticket.id)
+            data = ticket.wait()
+            self.call_take(data)
+
+    def try_to_stop(self, to_stop):
+        """
+        Attempt to stop TO_STOP resources by closing Resalloc tickets.  Not all
+        the resources may be closed at this time.
+        """
+        self.log.info("Trying to stop %s resources", to_stop)
+        stopped = 0
+        for ticket_id in self.tickets:
+            if stopped >= to_stop:
+                break
+
+            ticket = Ticket(ticket_id, connection=self.conn)
+            data = ticket.collect()
+            if not self.call_release(data):
+                self.log.debug("Can't release %s", ticket.id)
+                continue
+
+            self.log.debug("Closing ticket %s", ticket.id)
+            ticket.close()
+            self.tickets.remove(ticket_id)
+            stopped += 1
+
+
+    def loop(self):
+        """
+        Periodically query the ideal number of builders, and attempt to converge
+        to the ideal state.
+        """
+        while True:
+            start = time.time()
+            todo = self.call_converge_to() - len(self.tickets)
```

Comment:
We probably need a check here so that number of tickets opened should not be above maximum number of machines allowed in a pool.

Reply:
The set of resources may span across several pools so this is a hard task to do ... can we implement this later?

Reply:
Yes.

Reply:
If the `converge-to` hook returns a number greater than number of running machines, the `recycle` hook would never be called and all the running machines would be in an unusable state.

Reply:
The `self.recycle()` is called everytime.

Reply:
[This code](https://github.com/praiskup/resalloc/blob/93de07731dd7ad53bbedfd4878d2d3156e783178/agentspawner/daemon.py#L68-L72) synchronously waits for tickets, so if the `converge-to` hook returns a number greater than running nodes, `recyle` hook would never be called.

Reply:
Indeed -> it is better to perform the preparation async.  Noted.  I think I'm going to start using python-copr-common logic for dispatcher/background worker which depends on Redis, is that OK?  While depending on Redis, I think we can save the list of opened tickets there, too.

Reply:
@kdudka The hub and resalloc spawner daemon should be running on the same node. Would it cause any problems to have a dependency on redis?

Reply:
Redis can be hosted in a different container.  We just need the runtime dependency on redis client (python3-redis).